Modules/events/event.py

```diff
@@ -0,0 +1,181 @@
+# return this value to consume an event
+CANCEL_EVENT = object()
+
+
+class Event:
+    """
+        Defines an event.
+
+        an event is a subscribe-able object that will call its subscribers when it is invoked.
+
+    Examples:
+
+         an event can also be defined without a function definition
+         >>> my_other_event = Event("my_other_event")
+         >>> "my_other_event" in Event.events
+         True
+
+    Notes:
+        - Events will need to be invoked by something.
+
+        - Event subscribers must be async functions, and should be tested to work with
+            the event's signature (variable)
+
+        - exceptions raised by subscribers will not stop an event from being emitted
+
+        - the return value of decorated events are discarded.
+
+    See Also:
+        :meth:`subscribe`
+
+
+    """
+    events: Dict[str, 'Event'] = {}
+    """Registry mapping event names to their subscribers"""
+
+    def __init__(self, name: str):
+        """
+        Defines a new Event
+
+        Args:
+            name (str): name of the new event
+        """
+        if not isinstance(name, str):
+            raise TypeError(f"expected type str got {type(name)}")
+        self.name: str = name
+        """name of this event"""
+        self.subscribers: Subscriptions = []
+        """This events subscribers"""
+
+        Event._register(self)
+
+    async def _emit(self, *args, **kwargs):
+        """
+        async generator that does the actual event emission
+
+        Args:
+            *args ():
+            **kwargs ():
+
+        Yields:
+            return value of each subscriber
+        """
+        for subscriber in self.subscribers:
+            if isinstance(subscriber, staticmethod):
+                subscriber = subscriber.__func__
+            log.debug(f"calling subscriber {subscriber}...")
+            result = await subscriber(*args, **kwargs)
+            if result is CANCEL_EVENT:
+                return  # cancel the generator
+
+            yield result
+
+    async def emit(self, *args: Any, **kwargs: Any) -> List[Any]:
+        """
+        Emit the event to all subscribers
+
+        Args:
+            *args (tuple): tuple of positional arguments to subscribers
+            **kwargs (dict): dict of keyword arguments to provide to subscribers
+
+        Returns:
+
+        """
+        outcomes = [outcome async for outcome in self._emit(*args, **kwargs)]
+        log.debug(f"outcomes = {outcomes}")
+        return outcomes
+
+    @classmethod
+    def _register(cls, event: 'Event'):
+        """
+        Registers a new event
+
+        Args:
+            event(Event): Event to register
+
+        Raises:
+            ValueError: attempted to register an event has already been registered
+        """
+        if event.name in cls.events:
+            raise ValueError(f"name {event.name} is already registered as an event")
+        cls.events[event.name] = event
```

Comment:
Is `Event.events` used at all still?

Reply:
```py

    @classmethod
    def _register(cls, event: 'Event'):
        """
        Registers a new event

        Args:
            event(Event): Event to register

        Raises:
            ValueError: attempted to register an event has already been registered
        """
        if event.name in cls.events:
            raise ValueError(f"name {event.name} is already registered as an event")
        cls.events[event.name] = event

```
yes. I believe i use this to keep track of the actually registered events...
now that i think about it this should be made private